homeassistant/components/motionblinds_ble/cover.py

```diff
@@ -0,0 +1,717 @@
+        device_class=CoverDeviceClass.BLIND
+    ),
+    MotionBlindType.DOUBLE_ROLLER.value: MotionCoverEntityDescription(),
+    MotionBlindType.CURTAIN.value: MotionCoverEntityDescription(
+        device_class=CoverDeviceClass.CURTAIN
+    ),
+    MotionBlindType.VERTICAL.value: MotionCoverEntityDescription(
+        device_class=CoverDeviceClass.CURTAIN, icon=ICON_VERTICAL_BLIND
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up blind based on a config entry."""
+
+    blind_type = BLIND_TO_ENTITY_TYPE[entry.data[CONF_BLIND_TYPE]]
+    blind = blind_type(entry)
+
+    hass.data[DOMAIN][entry.entry_id] = blind
+
+    async_add_entities([blind])
+
+
+class GenericBlind(CoverEntity):
+    """Representation of a blind."""
+
+    device_address: str
+    device_rssi: int | None = None
+    _device: MotionDevice
+    _attr_connection_type: MotionConnectionType = MotionConnectionType.DISCONNECTED
+    _running_type: MotionRunningType | None = None
+
+    _use_status_position_update_ui: bool = False
+
+    _battery_callback: Callable[[int | None], None] | None = None
+    _speed_callback: Callable[[MotionSpeedLevel | None], None] | None = None
+    _connection_callback: Callable[[MotionConnectionType | None], None] | None = None
+    _signal_strength_callback: Callable[[int | None], None] | None = None
+
+    def __init__(self, entry: ConfigEntry) -> None:
+        """Initialize the blind."""
+        _LOGGER.debug(
+            "(%s) Setting up %s cover entity (%s)",
+            entry.data[CONF_MAC_CODE],
+            entry.data[CONF_BLIND_TYPE],
+            BLIND_TO_ENTITY_TYPE[entry.data[CONF_BLIND_TYPE]].__name__,
+        )
+        super().__init__()
+        self.entity_description = COVER_TYPES[entry.data[CONF_BLIND_TYPE]]
+        self.config_entry: ConfigEntry = entry
+        self.device_address: str = entry.data[CONF_ADDRESS]
+        self._attr_name: str = ENTITY_NAME.format(mac_code=entry.data[CONF_MAC_CODE])
+        self._attr_unique_id: str = entry.data[CONF_ADDRESS]
+        self._attr_device_info: DeviceInfo = DeviceInfo(
+            identifiers={(DOMAIN, entry.data[CONF_MAC_CODE])},
+            manufacturer=MANUFACTURER,
+            name=self._attr_name,
+        )
+        self._attr_is_closed: bool | None = None
+        self._attr_is_opening: bool | None = None
+        self._attr_is_closing: bool | None = None
+        self._attr_should_poll: bool = False
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added."""
+        ble_device = (
+            async_ble_device_from_address(self.hass, self.device_address)
+            if self.device_address
+            else None
+        )
+        self._device = MotionDevice(
+            self.device_address, ble_device, device_name=self._attr_name
+        )
```

Comment:
Actually I think this logic should definetely be moved into `async_setup_entry` and the `MotionDevice` instance should be stored in `hass.data`. That way the other platforms can access the `MotionDevice` directly and register their callbacks without depending on the cover entity.

Reply:
The way the platforms interact I think it makes more sense to have one entity that is responsible for the `MotionDevice`. Currently the `cover` entity registers the callbacks with the `MotionDevice`:
```
        self._device.register_running_callback(self.async_update_running)
        self._device.register_position_callback(self.async_update_position)
        self._device.register_connection_callback(self.async_update_connection)
        self._device.register_status_callback(self.async_update_status)
```
The status callback, for instance, includes information on position, tilt, battery, speed level and whether or not the end positions are set. This information is processed within the `cover` entity and then passed on through callbacks within the `cover` entity to platforms like `sensor` which will be added after this. If I were to move this out of that class, there is far less control over this. Calibration, for example, is not something that is reported by the blinds but should be deduced by the application based on end position info and blind type (where blind type is not known by the `MotionDevice` but set within HA), therefore the `cover` entity has various classes that contain logic dealing with information from `MotionDevice` and combine it with information set in HA such as the blind type to decide on the right information that gets passed to other platforms like the calibration sensor, connection sensor etc. That's why the `cover` entity has these callbacks and is passed through `hass.data`.

Reply:
Personally I'd be okay with this, but I'm pretty sure that a maintainer will sooner or later tell you that:

1. One entity should not depend on another when not absolutely necessary, which is not the case here
2. The logic needed to decide what information gets passed should be moved to the library

The former can be done in the way I described above and the latter could be realized by simply passing the `BlindType` to the library when you initialize the `MotionDevice` object and defining the `BlindType` enum inside the library.

I don't think this would need that much of refactoring while allowing the entities to be completely independent from each other and reducing the size of your PR.

Reply:
I currently decorate several functions with a method that can be overridden by child classes, for instance `run_command_function`. This function checks the `MotionCalibrationType`, which is a state that is saved in the `cover`, and does something depending on it's state. The reason I saved this in the cover is because it is not reported by the motor meaning I don't think it makes sense to put it in the `MotionDevice`. Instead, the calibration type is deduced from the end positions and whether or not the motor has been given a command from within Home Assistant. For instance, when no end positions are set for a *curtain blind specifically*  and the motor has been given a command to run to a position or go up/down, then the motor will automatically calibrate (which will not be reported by the motor). The behavior for this is different for a vertical blind, as it cannot simply be calibrated by running it to a position but requires an interface and user intervention. Therefore, these things are handled within the `cover` as they depend on a state that is not reported by the blind but inferred from HA commands. On the other hand, this information is also used by the calibration sensor (which will be added in the next PR) to show the calibration status.

Given the above I don't really see a way in which I can restructure it to not make the other entities depend on the `cover` as it contains this logic that is dependent on commands sent from Home Assistant.

Reply:
> This function checks the MotionCalibrationType, which is a state that is saved in the cover, and does something depending on it's state.

> the calibration type is deduced from the end positions and whether or not the motor has been given a command

> it tells the user to do so using an exception

> they depend on a state that is not reported by the blind but inferred from HA commands

All these things are able to be moved into the library. I completely understand your argument that the library's scope currently is set so that such things that depend on user interaction are not implemented in there. I'm just not sure that this will be enough for the maintainers to accept all this logic being implemented inside the entity or even inside the integration. Ideally they want you to simply forward method calls to your library client object and set some attributes when an update is received.

Also imagine some other project wants to use your library in another context. They would have to reimplement all this logic about the calibration and state assumption in their own project.

Reply:
That's true. I'll try to move everything to the library though that probably needs quite a bit of refactoring

Reply:
I refactored a lot of the things such as calibration and what direction the blind is moving in into the library and the integration is now about 2/3 the size of the original!